lib/block-supports/block-style-variations.php

```diff
@@ -79,31 +79,63 @@ function gutenberg_render_block_style_variation_support_styles( $parsed_block )
 		return $parsed_block;
 	}
 
-	$config = array(
-		'version' => WP_Theme_JSON_Gutenberg::LATEST_SCHEMA,
-		'styles'  => $variation_data,
-	);
-
 	$class_name         = gutenberg_get_block_style_variation_class_name( $parsed_block, $variation );
 	$updated_class_name = $parsed_block['attrs']['className'] . " $class_name";
+	$variation_instance = substr( $class_name, 9 );
+	$class_name         = ".$class_name";
+
+	/*
+	 * To support blocks with more complex selectors, that can apply styles to
+	 * inner markup or even different inner elements depending on the feature,
+	 * the variation style data needs to be manipulated some. If the root variation
+	 * styles are moved under a variations property they will generate as desired.
```

Comment:
I was trying to avoid writing an essay here. Would linking to the issue and PR help?

Reply:
I have a better idea on how best to explain this bit of the code now. I'll update the inline comment shortly.

For the discussion on this PR though:

The crux of the issue is that blocks can define their own selectors to direct certain styles to inner parts of the blocks markup. For example, the inner `img` element for an Image block needs to get the border styles whereas the other styles might still belong on the outer `figure` wrapper for the block.

Previously, the variation's data was treated as its own "theme.json partial". The problem here is that root styles for theme.json don't have the ability to be directed to inner parts as that's what the `styles.elements` and `styles.blocks` properties do.

The fix in this PR essentially moves the variation's root styles to under the `styles.blocks` property so that the block's custom selectors can be honored.

Reply:
Done https://github.com/WordPress/gutenberg/pull/62125/commits/2f33a3d130d8a84f4659c6f94e88abc2f9063430